src/java.base/windows/native/libnio/ch/UnixDomainSockets.c

```diff
@@ -84,21 +91,73 @@ jint unixSocketAddressToSockaddr(JNIEnv *env, jbyteArray addr, struct sockaddr_u
     return ret;
 }
 
+static int cmpGuid(GUID *g1, GUID *g2) {
+    if (g1->Data1 != g2->Data1)
+        return 0;
+    if (g1->Data2 != g2->Data2)
+        return 0;
+    if (g1->Data3 != g2->Data3)
+        return 0;
+    return !memcmp(g1->Data4, g2->Data4, 8);
+}
+
+static WSAPROTOCOL_INFO provider;
+
 JNIEXPORT jboolean JNICALL
 Java_sun_nio_ch_UnixDomainSockets_socketSupported(JNIEnv *env, jclass cl)
 {
-    SOCKET s = socket(PF_UNIX, SOCK_STREAM, 0);
-    if (s == INVALID_SOCKET) {
-        return JNI_FALSE;
+    WSAPROTOCOL_INFO info[5]; // if not large enough, a buffer is malloc'd
+    LPWSAPROTOCOL_INFO infoPtr = &info[0];
+    DWORD len = sizeof(info);
+    jboolean found;
+
+    /*
+     * First locate the Microsoft AF_UNIX Winsock provider
+     */
+    int result = WSAEnumProtocols(0, infoPtr, &len);
+    if (result == SOCKET_ERROR) {
+        if (GetLastError() == WSAENOBUFS) {
+            infoPtr = (LPWSAPROTOCOL_INFO)malloc(len);
+            result = WSAEnumProtocols(0, infoPtr, &len);
+            if (result == SOCKET_ERROR) {
+                free(infoPtr);
+                return JNI_FALSE;
+            }
+        } else {
+            return JNI_FALSE;
+        }
+    }
+    found = JNI_FALSE;
+    for (int i=0; i<result;  i++) {
+        if (infoPtr[i].iAddressFamily != AF_UNIX)
+            continue;
+        GUID g = infoPtr[i].ProviderId;
+        if (cmpGuid(&g, &MS_PROVIDER_ID)) {
+            found = JNI_TRUE;
+            provider = infoPtr[i];
+            break;
+        }
+    }
```

Comment:
The control flow can be simplified to avoid having continue and break in the loop.

Reply:
Okay